lapack/gonum/dgesc2.go

```diff
@@ -0,0 +1,83 @@
+// Copyright ©2021 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+// Dgesc2 solves a system of linear equations
+//  A * X = scale * RHS
+// with a general N-by-N matrix A using the LU factorization with
+// complete pivoting computed by Dgetc2. The result is placed in
+// rhs on exit.
+//
+// Dgesc2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgesc2(n int, a []float64, lda int, rhs []float64, ipiv, jpiv []int) (scale float64) {
+	switch {
+	case n < 0:
+		panic(nLT0)
+	case lda < max(1, n):
+		panic(badLdA)
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		return 0
+	}
+
+	switch {
+	case len(a) < (n-1)*lda+n:
+		panic(shortA)
+	case len(rhs) < n:
+		panic(shortRHS)
+	case len(ipiv) != n:
+		panic(badLenIpiv)
+	case len(jpiv) != n:
+		panic(badLenJpiv)
+	}
+
+	const smlnum = dlamchS / dlamchP
+	if len(a) < (n-1)*lda+n {
+		panic(shortA)
+	}
+
+	// Apply permutations ipiv to RHS.
+	impl.Dlaswp(1, rhs, 1, 0, n-1, ipiv[:n], 1)
```

Comment:
@kortschak or @vladimir-ch , I went back to `Dgesc2` to see how I used `Dlaswp` and this is confusing me some. The reference material does `CALL dlaswp( 1, rhs, lda, 1, n-1, ipiv, 1 )`, which according to its documentations, does row swap for `k2-k1+1` rows, that is **n-1** rows. Are we not swapping **n** rows here? Or am I misunderstanding?

Reply:
This was discussed in https://github.com/gonum/gonum/pull/1652#discussion_r646106943. The reference documents IPIV and JPIV to have length n, DGETC2 sets n elements but the only users of the PIV arrays, DGESC2 and DLATDF, refer only to the first n-1 elements. That's because n-1 swaps are enough to represent the permutation matrix from DGETC2 and from the way PIV arrays are constructed the last element must be n-1. DGESC2 uses this fact and does not call DLASWP with the last no-op swap.

I see three options:
1. Keep what we have: ipiv and jpiv have length n, Dgetc2 sets all elements, Dgesc2 and Dlatdf pass all n elements to Dlaswp, Dlaswp will not do anything for the last swap.
2. Do what the reference does: ipiv and jpiv have length n, Dgetc2 sets all elements but Dgesc2 and Dlatdf pass only n-1 elements to Dlaswp
3. ipiv and jpiv have length n-1, Dgetc2 sets all elements, Dgesc2 and Dlatdf pass all n-1 elements to Dlaswp.

I think what the reference does is the least clear of these options (set n elements but then use only n-1 elements for no documented, albeit logical, reason). Option 1 or option 3 are both fine, option 1 diverges less from the reference.

Reply:
Thank you for clearing it up. I remember you and Dan talked about it but I didn't really understand what the fuss was all about. I do now, crystal clear explanation, btw!